### PR TITLE
initial commit for treatments implementation into main chart

### PR DESCRIPTION
**Pull request by paulplant**
- stored/synced treatments will be displayed on the main chart
- carbs will be shown as a floating circle aligned with closest glucose point values
- bolus will be shown as a triangle. Micro-bolus threshold can be configured in settings (default: 1.0U). Micro-bolus will then be shown as small triangle with no labels
- option in settings to show/hide treatments on the chart (default is to show)
- changes to TreatmentType structure
- functions added to treatmentEntryAccessor
- new userdefault keys added to store new user settings
- a new chartpoint class extension has been added. This is modified from Loopkit
- correction to the NSResponse to correctly filter exercise treatments
- initial translations implemented

### CODE REVIEW

**Review by JohanDegraeve — CHANGES_REQUESTED** [review]
can you now see my review comments

File: xdrip/Constants/ConstantsGlucoseChart.swift [hunk 1]
```diff
@@ -77,12 +77,66 @@ enum ConstantsGlucoseChart {
     /// glucose colors - for values between highMarkValue and urgentHighMarkValue or between urgentLowMarkValue and lowMarkValue
     static let glucoseNotUrgentRangeColor = UIColor.yellow
     
-    /// calibration circle color (inside circle)
-    static let calibrationInsideColor = UIColor.red
+    /// diameter of the circle for blood glucose readings. The more hours on the chart, the smaller the circles should be
+    static let glucoseCircleDiameter3h: CGFloat = 7
+    static let glucoseCircleDiameter6h: CGFloat = 6
```
[the remark this answers is not in the document]
Comment (paulplant) [reply xdrip/Constants/ConstantsGlucoseChart.swift:82]: Done

File: xdrip/Constants/ConstantsGlucoseChart.swift [hunk 2]
```diff
@@ -77,12 +77,66 @@ enum ConstantsGlucoseChart {
...
+    static let glucoseCircleDiameter24h: CGFloat = 4
+    
+    /// calibration circle color + scaling
+    static let calibrationCircleColorInner = UIColor.red
+    static let calibrationCircleColorOuter = UIColor.white
+    static let calibrationCircleScaleOuter: CGFloat = 1.9
+    static let calibrationCircleScaleInner: CGFloat = 1.4
+    
+    /// bolus treament chart colour + y-axis scaling + offset.
+    static let bolusTreatmentColor = UIColor.systemBlue
```
[the remark this answers is not in the document]
Comment (paulplant) [reply xdrip/Constants/ConstantsGlucoseChart.swift:93]: Comment corrected

File: xdrip/Core Data/classes/ChartPointsScatterDownTrianglesLayer.swift [hunk 3]
```diff
@@ -0,0 +1,55 @@
+//
+//  ChartPointsScatterDownTrianglesLayer.swift
```
[the remark this answers is not in the document]
Comment (paulplant) [reply xdrip/Core Data/classes/ChartPointsScatterDownTrianglesLayer.swift:2]: Done

File: xdrip/Core Data/classes/TreatmentEntry+CoreDataClass.swift [hunk 4]
```diff
@@ -72,6 +72,38 @@ import CoreData
...
+    
+    /// return the y-axis offset for the treatment type (as set in ConstantsGlucoseChart.swift) - if it isn't required for the treatment type, it should return 0
+    public func chartPointYAxisOffset() -> Double {
+        
+        switch self {
+            
+        case .Insulin:
+            return ConstantsGlucoseChart.bolusTreatmentChartPointYAxisOffsetInMgDl.mgdlToMmol(mgdl: UserDefaults.standard.bloodGlucoseUnitIsMgDl)
+            
+            // if no offset is defined or  this treatment type, just return zero offset
```
[the remark this answers is not in the document]
Comment (paulplant) [reply xdrip/Core Data/classes/TreatmentEntry+CoreDataClass.swift:84]: Corrected. It should have read "if no offset is defined or needed for this treatment type...."

File: xdrip/Extensions/ChartPoint.swift [hunk 5]
```diff
@@ -20,6 +20,36 @@ extension ChartPoint {
             )
 
     }
+    
+    // the chartpoints defined for bolus treatment entries are abolute-positioned in the chart and need to be scaled to fit the y-axis values of the glucose chart points (and therefore avoid needing a secondary axis). The offset from the bottom of the chart and the scale is pulled from the Treatment Type.
```
[the remark this answers is not in the document]
Comment (paulplant) [reply xdrip/Extensions/ChartPoint.swift:24]: Done

File: xdrip/Managers/Charts/GlucoseChartManager.swift [hunk 6]
```diff
@@ -16,6 +16,12 @@ public class GlucoseChartManager {
     /// - maximumValueInGlucoseChartPoints = the largest x value (ie the highest Glucose value) considering the three arrays together
     typealias GlucoseChartPointsType = (urgentRange: [ChartPoint], inRange: [ChartPoint], notUrgentRange: [ChartPoint], firstGlucoseChartPoint: ChartPoint?, lastGlucoseChartPoint: ChartPoint?, maximumValueInGlucoseChartPoints: Double?)
     
+    /// to hold the treatment chartpoints
+    /// - smallBolus = bolus values below themicro-bolus threshold (usually around 1.0U or less)
```
[the remark this answers is not in the document]
Comment (paulplant) [reply xdrip/Managers/Charts/GlucoseChartManager.swift:20]: Done!

File: xdrip/Managers/Charts/GlucoseChartManager.swift [hunk 7]
```diff
@@ -29,14 +35,26 @@ public class GlucoseChartManager {
 
     /// CalibrationPoints to be shown on chart
     private var calibrationChartPoints = [ChartPoint]()
+        
+    private var treatmentChartPoints: TreatmentChartPointsType = ([ChartPoint](), [ChartPoint](), [ChartPoint](), [ChartPoint](), [ChartPoint](), [ChartPoint]())
```
[the remark this answers is not in the document]
Comment (paulplant) [reply xdrip/Managers/Charts/GlucoseChartManager.swift:39]: Done

File: xdrip/Managers/Charts/GlucoseChartManager.swift [hunk 8]
```diff
@@ -29,14 +35,26 @@ public class GlucoseChartManager {
...
+        
+    /// smallBolusTreatmentChartPoints to be shown on chart
+    private var smallBolusTreatmentChartPoints = [ChartPoint]()
```
[the remark this answers is not in the document]
Comment (paulplant) [reply xdrip/Managers/Charts/GlucoseChartManager.swift:42]: Done

File: xdrip/Managers/Charts/GlucoseChartManager.swift [hunk 9]
```diff
@@ -854,6 +1013,259 @@ public class GlucoseChartManager {
...
+        let minutesEitherSide: Double = 10
+        let minutesEitherSideExtended: Double = 70
+        
+        var lowerValueToUse: Double = 0
+        var upperValueToUse: Double = 0
+        
+        // in normal circumstances there will always be CGM data but in the unlikely event that there is no data point to calculate 1 hour before and after, we'll just pin it to the user's target objective which should be in the middle of their desired range
+        let bgReadingsBefore = bgReadingsAccessor.getBgReadings(from: treatmentDate.addingTimeInterval(-minutesEitherSide * 60), to: treatmentDate, on: managedObjectContext)
+        
+        if bgReadingsBefore.count > 0 {
```
[the remark this answers is not in the document]
Comment (paulplant) [reply xdrip/Managers/Charts/GlucoseChartManager.swift:1113]: Done

File: xdrip/Managers/Charts/GlucoseChartManager.swift [hunk 10]
```diff
@@ -854,6 +1013,259 @@ public class GlucoseChartManager {
...
+            
+            lowerValueToUse = bgReadingsBefore.last!.calculatedValue
+            
+        } else {
+            
+            let bgReadingsBeforeExtended = bgReadingsAccessor.getBgReadings(from: treatmentDate.addingTimeInterval(-minutesEitherSideExtended * 60), to: treatmentDate, on: managedObjectContext)
```
[the remark this answers is not in the document]
Comment (paulplant) [reply xdrip/Managers/Charts/GlucoseChartManager.swift:1119]: Done!

File: xdrip/Managers/Charts/GlucoseChartManager.swift [hunk 11]
```diff
@@ -854,6 +1013,259 @@ public class GlucoseChartManager {
...
+        var mediumBolusTreatmentEntryChartPoints = [ChartPoint]()
+        var smallCarbsTreatmentEntryChartPoints = [ChartPoint]()
+        var mediumCarbsTreatmentEntryChartPoints = [ChartPoint]()
+        var largeCarbsTreatmentEntryChartPoints = [ChartPoint]()
+        var veryLargeCarbsTreatmentEntryChartPoints = [ChartPoint]()
+        
+        managedObjectContext.performAndWait {
+            
+            // cycle through the treatment entries and append the chart points
+            for treatmentEntry in treatmentEntries {
```
[the remark this answers is not in the document]
Comment (paulplant) [reply xdrip/Managers/Charts/GlucoseChartManager.swift:1042]: Done

**Comment by paulplant** [comment]
Decimal point issue fixed and committed @LiroyvH,  thanks!

**Review by JohanDegraeve — APPROVED** [review]

**Comment by tromanisagit** [comment]
I got the 0 entry on the xDrip4iO5 chart, but only once, other attempts worked.  Delete works too. This was pre change, so will build again and test.  

My pref is to have some form of iob stat showing too, to track stacking of treatments.  Pref would be xDrip4iO5 calculated rather than Nightscout as offline support.

**Comment by LiroyvH** [comment]
@paulplant Danke sir!
Before I forget to log it (which I kinda did for the past 2 weeks 😂😂): this problem is confirmed solved in the micro-bolus section, but still exists in the treatments window.

**Comment by paulplant** [comment]
Liro

> @paulplant Danke sir!
> Before I forget to log it (which I kinda did for the past 2 weeks 😂😂): this problem is confirmed solved in the micro-bolus section, but still exists in the treatments window.

Fixed in #341 @LiroyvH , we'll get it merged into develop shortly.

This PR has actually actually been merged in by @JohanDegraeve so not sure why it's showing as Open still? I'll close it to prevent confusion.

**Pull request closed without merging**